flytekit/tools/ignore.py

```diff
@@ -1,17 +1,128 @@
 import os
+import re
 import subprocess
+import sys
 import tarfile as _tarfile
 from abc import ABC, abstractmethod
 from fnmatch import fnmatch
 from pathlib import Path
 from shutil import which
 from typing import Dict, List, Optional, Type
 
-from docker.utils.build import PatternMatcher
-
 from flytekit.loggers import logger
 
+IS_WINDOWS_PLATFORM = sys.platform == "win32"
+
 STANDARD_IGNORE_PATTERNS = ["*.pyc", ".cache", ".cache/*", "__pycache__", "**/__pycache__"]
+_SEP = re.compile("/|\\\\") if IS_WINDOWS_PLATFORM else re.compile("/")
+
+
+def split_path(p):
+    return [pt for pt in re.split(_SEP, p) if pt and pt != "."]
+
+
+def normalize_slashes(p):
+    if IS_WINDOWS_PLATFORM:
+        return "/".join(split_path(p))
+    return p
+
+
+# It's copied from docker module to reduce flytekit module size.
+# reference: https://github.com/docker/docker-py/blob/9ad4bddc9ee23f3646f256280a21ef86274e39bc/docker/utils/build.py#L159
```

Comment:
When copying over code from a Apache 2.0 project, we have to copy over their license into our repo:

https://github.com/docker/docker-py/blob/9ad4bddc9ee23f3646f256280a21ef86274e39bc/LICENSE

There are a few ways to do this. The easy way is to have the vendored code in its own file, `flytekit/tools/docker.py` and place `PatternMatcher` there. Then copy and paste the license at the top of the file. For example: https://github.com/scipy/scipy/blob/main/scipy/_lib/_pep440.py

Reply:
Very nice advice, thank you so much!